### PR TITLE
Tooltips

### PR DESCRIPTION
**Pull request by willmcgugan**
- Adds Widget.tooltip
- Added Region.inflect
- Added `constrain: inflect;` Currently undocumented, as it is not stable.

https://github.com/Textualize/textual/assets/554369/a0b8bc84-4ca2-4ff0-930b-0696e9829b7e

### CODE REVIEW

File: src/textual/_doc.py [hunk 1]
```diff
@@ -115,7 +119,11 @@ async def auto_pilot(pilot: Pilot) -> None:
             result = run_before(pilot)
             if inspect.isawaitable(result):
                 await result
+        await pilot.pause()
         await pilot.press(*press)
+        if hover:
+            await pilot.hover(hover)
+            await pilot.pause(0.5)
```
Comment (rodrigogiraoserrao) [inline src/textual/_doc.py:126]: For the purposes of testing and auto-pilots, isn't there something else we can call to trigger the tooltip immediately? Waiting half a second for a tooltip to show up in CI can compound to a lot of time.

File: src/textual/geometry.py [hunk 2]
```diff
@@ -905,6 +905,47 @@ def translate_inside(
...
+                        │    ┌──────────┐
+                             │          │
+                        │    │  Result  │
+                             │          │
+                        │    └──────────┘
+        ```
+
+        Args:
+            x_axis: +1 to infect in the positive direction, -1 to inflect in the negative direction.
+            y_axis: +1 to infect in the positive direction, -1 to inflect in the negative direction.
```
Comment (rodrigogiraoserrao) [inline src/textual/geometry.py:935]: 🤧 😷 

```suggestion
            x_axis: +1 to inflect in the positive direction, -1 to inflect in the negative direction.
            y_axis: +1 to inflect in the positive direction, -1 to inflect in the negative direction.
```

File: src/textual/geometry.py [hunk 3]
```diff
@@ -905,6 +905,47 @@ def translate_inside(
...
         )
 
+    def inflect(
+        self, x_axis: int = +1, y_axis: int = +1, margin: Spacing | None = None
+    ) -> Region:
+        """Inflect a region around one or both axis.
+
+        The `x_axis` and `y_axis` parameters define which direction to move the region.
+        A positive value will move the region right or down, a negative value will move
+        the region left or up.
```
Comment (rodrigogiraoserrao) [inline src/textual/geometry.py:915]: Should we specify that the value `0` does _not_ inflect the given axis?

File: src/textual/screen.py [hunk 4]
```diff
@@ -169,6 +176,18 @@ def _update_timer(self) -> Timer:
             )
         return self.__update_timer
 
+    @property
+    def layers(self) -> tuple[str, ...]:
+        """Layers of from parent.
```
Comment (rodrigogiraoserrao) [inline src/textual/screen.py:181]: Not sure what you meant here?

```suggestion
        """Layers of names from parent.
```

```suggestion
        """Layers from parent.
```

File: src/textual/widget.py [hunk 5]
```diff
@@ -367,7 +370,7 @@ def __init__(
     """Show a horizontal scrollbar?"""
 
     show_horizontal_scrollbar: Reactive[bool] = Reactive(False, layout=True)
-    """SHow a vertical scrollbar?"""
+    """Show a vertical scrollbar?"""
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:373]:
```suggestion
    """Show a horizontal scrollbar?"""
```

File: src/textual/widget.py [hunk 6]
```diff
@@ -3103,8 +3120,16 @@ async def _on_compose(self) -> None:
...
+            self._extend_compose(widgets)
             await self.mount(*widgets)
 
+    def _extend_compose(self, widgets: list[Widget]) -> None:
+        """Hook to extend composed widgets.
+
+        Args:
+            widgets: Widgets to be mounted.
+        """
+
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:3132]: I don't get the overarching idea of adding `_extend_compose`. 😞

**Review by rodrigogiraoserrao — CHANGES_REQUESTED** [review]
Should we have a snapshot test of a tooltip?

File: src/textual/_doc.py (hunk 1 above)
Comment (willmcgugan) [reply src/textual/_doc.py:126]: It is an issue, but anything else wouldn't be a fair test.

I did start implementing a virtual clock of sorts, that could run at 10x speed when testing. But that didn't work out.

File: src/textual/widget.py (hunk 6 above)
Comment (willmcgugan) [reply src/textual/widget.py:3132]: It's to allow the Screen to add a Tooltip to the DOM, without requiring it to be explicitly added to `compose()`

File: src/textual/geometry.py (hunk 3 above)
Comment (willmcgugan) [reply src/textual/geometry.py:915]: I guess so.

**Comment by willmcgugan** [comment]
We don't have a way of snapshot testing tooltips. It requires that you move the (virtual) mouse over a button and wait for the timeout. Will revisit when we have updated the snapshot API to support that.

File: tests/snapshot_tests/snapshot_apps/scroll_to_center.py [hunk 7]
```diff
@@ -21,6 +21,8 @@ class MyApp(App[None]):
     }
     """
 
+    # AUTO_FOCUS = ""
+
```
Comment (rodrigogiraoserrao) [inline tests/snapshot_tests/snapshot_apps/scroll_to_center.py:25]:
```suggestion
```

**Review by rodrigogiraoserrao — APPROVED** [review]

**Pull request merged**